homeassistant/helpers/intent.py

```diff
@@ -663,120 +635,283 @@ def async_match_targets(  # noqa: C901
+    preferences: MatchTargetsPreferences,
+    candidates: list[MatchTargetsCandidate],
+    area_candidate_filter: Callable[[MatchTargetsCandidate, Collection[str]], bool],
+    areas_added: bool,
+) -> tuple[list[MatchTargetsCandidate], MatchTargetsResult | None, bool]:
+    """Handle single target constraint."""
+    if not constraints.single_target or len(candidates) <= 1:
+        return candidates, None, areas_added
+
+    # Find best match using preferences
+    if not (preferences.area_id or preferences.floor_id):
+        # No preferences
+        return (
+            candidates,
+            MatchTargetsResult(
                 False,
-                MatchFailedReason.NAME,
-                areas=targeted_areas or [],
-                floors=targeted_floors or [],
-            )
+                MatchFailedReason.MULTIPLE_TARGETS,
+                states=[c.state for c in candidates],
+            ),
+            areas_added,
+        )
 
-        candidates = final_candidates
+    if not areas_added:
+        ar = area_registry.async_get(hass)
+        dr = device_registry.async_get(hass)
+        _add_areas(ar, dr, candidates)
+        areas_added = True
 
-    if constraints.single_target and len(candidates) > 1:
-        # Find best match using preferences
-        if not (preferences.area_id or preferences.floor_id):
-            # No preferences
-            return MatchTargetsResult(
+    filtered_candidates: list[MatchTargetsCandidate] = candidates
+    if preferences.area_id:
+        # Filter by area
+        filtered_candidates = [
+            c for c in candidates if area_candidate_filter(c, {preferences.area_id})
+        ]
+
+    if (len(filtered_candidates) > 1) and preferences.floor_id:
+        # Filter by floor
+        filtered_candidates = [
+            c
+            for c in candidates
+            if c.area and (c.area.floor_id == preferences.floor_id)
+        ]
+
+    if len(filtered_candidates) != 1:
+        # Filtering could not restrict to a single target
+        return (
+            candidates,
+            MatchTargetsResult(
                 False,
                 MatchFailedReason.MULTIPLE_TARGETS,
                 states=[c.state for c in candidates],
-            )
+            ),
+            areas_added,
+        )
+
+    return filtered_candidates, None, areas_added
+
+
+@callback
+def async_match_targets(  # noqa: C901
+    hass: HomeAssistant,
+    constraints: MatchTargetsConstraints,
+    preferences: MatchTargetsPreferences | None = None,
+    states: list[State] | None = None,
+    area_candidate_filter: Callable[
+        [MatchTargetsCandidate, Collection[str]], bool
+    ] = _default_area_candidate_filter,
+) -> MatchTargetsResult:
+    """Match entities based on constraints in order to handle an intent."""
+    preferences = preferences or MatchTargetsPreferences()
+
+    candidates, result = _get_candidates_by_constraints(hass, constraints, states)
+    if result:
```

Comment:
```suggestion
    if result is not None:
```
Be specific, same in other places.